### PR TITLE
Merge app non app version of pbi dashboards

### PR DESCRIPTION
**Pull request by elic-eon**
### 🤔 Why?

No reason to keep both app and non-app version of dashboards

### 🤓 What?

- Merge app information into non-app version
- Use `poetry` 1.2 to generate lock file. `poetry lock --no-update`

### 🧪 Tested?

Unit tests. Compared before after, we have 8 assets with app enabled in our PowerBI, verified that we reduce 8 asset after the change.

### CODE REVIEW

File: metaphor/power_bi/extractor.py [hunk 1]
```diff
@@ -313,6 +327,9 @@ def _find_last_completed_refresh(
         iso_time = ".".join(refresh.endTime.split(".")[0:-1])
         return datetime.fromisoformat(iso_time).replace(tzinfo=timezone.utc)
 
+    def _get_original_id_from_url(self, url: str) -> str:
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:330]:
```suggestion
    def _get_dashboard_id_from_url(self, url: str) -> str:
```

File: metaphor/power_bi/extractor.py [hunk 2]
```diff
@@ -276,6 +277,19 @@ def map_wi_dashboards_to_dashboard(
             )
             self._dashboards[wi_dashboard.id] = dashboard
 
+    def merge_app_version_into_original_version(self):
+        for dashboard_id in list(self._dashboards.keys()):
+            dashboard = self._dashboards.get(dashboard_id)
+            if dashboard.dashboard_info.power_bi.app:
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:283]:
```suggestion
            if dashboard.dashboard_info.power_bi.app is not None:
```

File: metaphor/power_bi/extractor.py [hunk 3]
```diff
@@ -276,6 +277,19 @@ def map_wi_dashboards_to_dashboard(
...
+                original_id = self._get_original_id_from_url(
+                    dashboard.source_info.main_url
+                )
+                original_dashboard = self._dashboards.get(original_id)
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:287]: Might want to defensively check that `original_dashboard` is not None.

Comment (elic-eon) [reply metaphor/power_bi/extractor.py]: Sounds good.

Comment (elic-eon) [reply metaphor/power_bi/extractor.py]: Will keep app version if non-app version is not exist (shouldn't be possible)

File: metaphor/power_bi/extractor.py [hunk 4]
```diff
@@ -276,6 +277,26 @@ def map_wi_dashboards_to_dashboard(
             )
             self._dashboards[wi_dashboard.id] = dashboard
 
+    def merge_app_version_into_original_version(self):
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:280]:
```suggestion
    def dedupe_app_version_dashboards(self):
```

Comment (mars-lan) [reply metaphor/power_bi/extractor.py:280]: Also add a comment for posterity.

File: metaphor/power_bi/extractor.py [hunk 5]
```diff
@@ -276,6 +277,26 @@ def map_wi_dashboards_to_dashboard(
...
+            if dashboard.dashboard_info.power_bi.app is None:
+                # Non-app dashboard
+                continue
+
+            original_dashboard_id = self._get_dashboard_id_from_url(
+                dashboard.source_info.main_url
+            )
+            original_dashboard = self._dashboards.get(original_dashboard_id)
+            if original_dashboard is None:
+                # Cannot not found corresponding non-app dashboard
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:292]: `logger.warn`?

File: metaphor/power_bi/extractor.py [hunk 6]
```diff
@@ -313,6 +337,9 @@ def _find_last_completed_refresh(
         iso_time = ".".join(refresh.endTime.split(".")[0:-1])
         return datetime.fromisoformat(iso_time).replace(tzinfo=timezone.utc)
 
+    def _get_dashboard_id_from_url(self, url: str) -> str:
+        return url.rstrip("/").split("/")[-1]
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:341]: Let's use proper regex for this? Also you may wish to use [urllib.parse](https://docs.python.org/3/library/urllib.parse.html#module-urllib.parse) to extract the path portion.

File: metaphor/power_bi/extractor.py [hunk 7]
```diff
@@ -276,6 +277,29 @@ def map_wi_dashboards_to_dashboard(
...
+                continue
+
+            original_dashboard_id = self._get_dashboard_id_from_url(
+                dashboard.source_info.main_url
+            )
+            original_dashboard = self._dashboards.get(original_dashboard_id)
+            if original_dashboard is None:
+                # Cannot not found corresponding non-app dashboard
+                logger.warning(
+                    f"Non app version dashboard not found, id: {dashboard_id}"
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:294]:
```suggestion
                    f"Non-app version dashboard not found, id: {dashboard_id}"
```

File: metaphor/power_bi/extractor.py [hunk 8]
```diff
@@ -337,8 +339,19 @@ def _find_last_completed_refresh(
...
         return datetime.fromisoformat(iso_time).replace(tzinfo=timezone.utc)
 
-    def _get_dashboard_id_from_url(self, url: str) -> str:
-        return url.rstrip("/").split("/")[-1]
+    def _get_dashboard_id_from_url(self, url: str) -> Optional[str]:
+        path = parse.urlparse(url).path
+        print(path)
+        pattern = re.compile(
+            r"apps/"
+            r"([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/"
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:347]: This might be too specific to be robust over the long term. Let's use something like `[^/]+`?

Comment (elic-eon) [reply metaphor/power_bi/extractor.py:347]: Sounds good

Comment (elic-eon) [reply metaphor/power_bi/extractor.py]: Updated.

File: metaphor/power_bi/extractor.py [hunk 9]
```diff
@@ -276,6 +279,29 @@ def map_wi_dashboards_to_dashboard(
...
+            )
+            original_dashboard = self._dashboards.get(original_dashboard_id)
+            if original_dashboard is None:
+                # Cannot not found corresponding non-app dashboard
+                logger.warning(
+                    f"Non-app version dashboard not found, id: {original_dashboard_id}"
+                )
+                continue
+
+            original_dashboard.dashboard_info.power_bi.app = (
```
Comment (mars-lan) [inline metaphor/power_bi/extractor.py:300]: Let's also use the app version URL for the original dashboard.

**Review by mars-lan — APPROVED** [review]

**Pull request merged**